File: ios/Sources/TorchMath.swift
import Foundation

/// Pure helpers for torch math, kept free of AVFoundation/UIKit so they can be
/// unit-tested without a device or simulator.
public enum TorchMath {

    /// Clamps an optional `0.0...1.0` brightness into a valid `Float` level.
    ///
    /// `nil` means "no explicit level requested" and maps to full brightness.
    public static func clampLevel(_ level: Double?) -> Float {
        let value = Float(level ?? 1.0)
        return min(max(value, 0.0), 1.0)
    }

    /// Whether the requested level is effectively "maximum", in which case the
    /// caller should use `AVCaptureMaxAvailableTorchLevel` instead of passing
    /// `1.0` to `setTorchModeOn(level:)` (which can throw).
    public static func isMaxLevel(_ level: Float) -> Bool {
        return level >= 1.0
    }
}

File: android/src/main/java/TorchMath.kt
package com.lembryo.tauri.plugin.torchlight

import kotlin.math.roundToInt

/**
 * Pure helpers for the torch math, kept free of Android framework types so they
 * can be unit-tested on the host JVM.
 */
internal object TorchMath {

    /**
     * Maps a `0.0..1.0` brightness onto the discrete `1..maxLevel` strength range
     * used by [android.hardware.camera2.CameraManager.turnOnTorchWithStrengthLevel].
     *
     * The result is always clamped to a valid, on (`>= 1`) strength.
     */
    fun strengthFromLevel(level: Double, maxLevel: Int): Int {
        if (maxLevel <= 1) return 1
        val clamped = level.coerceIn(0.0, 1.0)
        val scaled = (clamped * maxLevel).roundToInt()
        return scaled.coerceIn(1, maxLevel)
    }

    /**
     * Whether a per-strength brightness level should be applied for the given
     * device maximum. Devices that report `maxLevel <= 1` have no brightness
     * control, so a plain on/off toggle must be used instead.
     */
    fun supportsBrightness(maxLevel: Int): Boolean = maxLevel > 1
}
